File: src/download.c
#include "download.h"
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>
#include <curl/curl.h>

#define LOCAL_SOURCE_PREFIX "file://"
#define LOCAL_SOURCE_PREFIX_LEN 7
#define NPM_SOURCE_PREFIX "npm:"
#define NPM_SOURCE_PREFIX_LEN 4

static bool curl_initialized = false;

int download_init(void) {
    if (curl_initialized) return 0;

    if (curl_global_init(CURL_GLOBAL_DEFAULT) != CURLE_OK) {
        log_error("Failed to initialize curl");
        return -1;
    }

    curl_initialized = true;
    return 0;
}

void download_cleanup(void) {
    if (curl_initialized) {
        curl_global_cleanup();
        curl_initialized = false;
    }
}

bool source_is_local(const char *source) {
    if (!source) return false;
    return strncmp(source, LOCAL_SOURCE_PREFIX, LOCAL_SOURCE_PREFIX_LEN) == 0;
}

const char *source_local_path(const char *source) {
    if (!source_is_local(source)) return NULL;
    return source + LOCAL_SOURCE_PREFIX_LEN;
}

bool source_is_npm(const char *source) {
    if (!source) return false;
    return strncmp(source, NPM_SOURCE_PREFIX, NPM_SOURCE_PREFIX_LEN) == 0;
}

const char *source_npm_after_prefix(const char *source) {
    if (!source_is_npm(source)) return NULL;
    return source + NPM_SOURCE_PREFIX_LEN;
}

void source_npm_split(const char *source, char **pkg_out, char **file_out) {
    if (pkg_out) *pkg_out = NULL;
    if (file_out) *file_out = NULL;
    const char *body = source_npm_after_prefix(source);
    if (!body) return;

    // Split on the LAST '#' so scoped packages (which contain no '#') and any
    // future quirks are tolerated. npm package names cannot contain '#' so
    // the first '#' is also fine — using last is just defensive.
    const char *hash = strchr(body, '#');
    if (!hash) {
        if (pkg_out) *pkg_out = str_dup(body);
        return;
    }
    size_t pkg_len = (size_t)(hash - body);
    if (pkg_out) {
        char *pkg = spm_malloc(pkg_len + 1);
        memcpy(pkg, body, pkg_len);
        pkg[pkg_len] = '\0';
        *pkg_out = pkg;
    }
    if (file_out && hash[1] != '\0') {
        *file_out = str_dup(hash + 1);
    }
}

// True if the user-supplied argument should be treated as a local-path skill
// rather than an owner/repo spec. We use leading-character cues that cannot
// occur in a valid GitHub owner name.
static bool looks_like_local_path(const char *spec) {
    if (!spec || !spec[0]) return false;
    if (strcmp(spec, ".") == 0) return true;
    if (spec[0] == '/') return true;
    if (spec[0] == '~' && spec[1] == '/') return true;
    if (spec[0] == '.' && spec[1] == '/') return true;
    if (spec[0] == '.' && spec[1] == '.' && spec[2] == '/') return true;
    return false;
}

// Resolve a user-supplied path to a "./<rel>" form rooted at the current
// working directory. Expands a leading "~/", canonicalises via realpath(),
// rejects paths outside the cwd. Returns NULL on error (already logged).
static char *canonicalize_local_path(const char *user_path) {
    if (!user_path || !user_path[0]) return NULL;

    char *expanded = NULL;
    if (user_path[0] == '~' && user_path[1] == '/') {
        char *home = get_home_dir();
        if (!home) {
            log_error("Cannot expand ~ (HOME not set)");
            return NULL;
        }
        size_t len = strlen(home) + 1 + strlen(user_path + 2) + 1;
        expanded = spm_malloc(len);
        snprintf(expanded, len, "%s/%s", home, user_path + 2);
        spm_free(home);
    } else {
        expanded = str_dup(user_path);
    }

    char *abs = realpath(expanded, NULL);
    spm_free(expanded);
    if (!abs) {
        log_error("Cannot resolve path: %s", user_path);
        return NULL;
    }

    char *cwd = getcwd(NULL, 0);
    if (!cwd) {
        log_error("Cannot get current directory");
        free(abs);
        return NULL;
    }

    size_t cwd_len = strlen(cwd);
    bool inside = strncmp(abs, cwd, cwd_len) == 0 &&
                  (abs[cwd_len] == '/' || abs[cwd_len] == '\0');

    if (!inside) {
        log_error("Local skill path is outside the project: %s", user_path);
        free(abs);
        free(cwd);
        return NULL;
    }

    const char *rel = abs + cwd_len;
    while (*rel == '/') rel++;

    char *result;
    if (*rel == '\0') {
        result = str_dup(".");
    } else {
        size_t len = 2 + strlen(rel) + 1;
        result = spm_malloc(len);
        snprintf(result, len, "./%s", rel);
    }

    free(abs);
    free(cwd);
    return result;
}

PackageSpec *package_spec_parse(const char *spec) {
    if (!spec) return NULL;

    // Detect local-path or file:// forms before treating as owner/repo.
    const char *local_input = NULL;
    if (source_is_local(spec)) {
        local_input = source_local_path(spec);
    } else if (looks_like_local_path(spec)) {
        local_input = spec;
    }

    if (local_input) {
        char *canonical = canonicalize_local_path(local_input);
        if (!canonical) return NULL;

        PackageSpec *ps = spm_malloc(sizeof(PackageSpec));
        ps->owner = NULL;
        ps->repo = NULL;
        ps->ref = NULL;
        ps->ref_explicit = false;
        ps->skill_in_spec = NULL;
        ps->is_local = true;
        ps->local_path = canonical;
        return ps;
    }

    PackageSpec *ps = spm_malloc(sizeof(PackageSpec));
    ps->owner = NULL;
    ps->repo = NULL;
    ps->ref = NULL;
    ps->ref_explicit = false;
    ps->skill_in_spec = NULL;
    ps->is_local = false;
    ps->local_path = NULL;

    // Make a working copy
    char *work = str_dup(spec);

    // Check for @ref suffix
    char *at = strchr(work, '@');
    if (at) {
        *at = '\0';
        ps->ref = str_dup(at + 1);
        ps->ref_explicit = true;
    } else {
        ps->ref = str_dup("main");
        ps->ref_explicit = false;
    }

    // Check for #skill suffix (after stripping @ref)
    char *hash = strchr(work, '#');
    if (hash) {
        *hash = '\0';
        if (hash[1] != '\0') {
            ps->skill_in_spec = str_dup(hash + 1);
        }
    }

    // Parse owner/repo
    char *slash = strchr(work, '/');
    if (!slash) {
        log_error("Invalid package spec: %s (expected owner/repo)", spec);
        spm_free(work);
        spm_free(ps->ref);
        spm_free(ps->skill_in_spec);
        spm_free(ps);
        return NULL;
    }

    *slash = '\0';
    ps->owner = str_dup(work);
    ps->repo = str_dup(slash + 1);

    spm_free(work);

    if (!ps->owner[0] || !ps->repo[0]) {
        log_error("Invalid package spec: %s (empty owner or repo)", spec);
        package_spec_free(ps);
        return NULL;
    }

    return ps;
}

void package_spec_free(PackageSpec *spec) {
    if (!spec) return;
    spm_free(spec->owner);
    spm_free(spec->repo);
    spm_free(spec->ref);
    spm_free(spec->skill_in_spec);
    spm_free(spec->local_path);
    spm_free(spec);
}

char *build_tarball_url(const PackageSpec *spec, RefKind kind) {
    if (!spec) return NULL;

    const char *kind_segment = (kind == REF_KIND_TAG) ? "tags" : "heads";
    const char *fmt = "https://github.com/%s/%s/archive/refs/%s/%s.tar.gz";

    size_t len = strlen(fmt) + strlen(spec->owner) + strlen(spec->repo)
                 + strlen(kind_segment) + strlen(spec->ref) + 1;
    char *url = spm_malloc(len);
    snprintf(url, len, fmt, spec->owner, spec->repo, kind_segment, spec->ref);

    return url;
}

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    FILE *fp = (FILE *)userp;
    return fwrite(contents, size, nmemb, fp);
}

// Internal: returns 0 on transport success and writes HTTP code to *out_http_code.
// Returns -1 on transport failure (curl error). Does not log on HTTP errors;
// caller decides what to do with the status.
static int download_file_internal(const char *url, const char *output_path, long *out_http_code) {
    if (!curl_initialized) {
        log_error("Curl not initialized");
        return -1;
    }

    CURL *curl = curl_easy_init();
    if (!curl) {
        log_error("Failed to create curl handle");
        return -1;
    }

    FILE *fp = fopen(output_path, "wb");
    if (!fp) {
        log_error("Cannot create file: %s", output_path);
        curl_easy_cleanup(curl);
        return -1;
    }

    log_debug("Downloading: %s", url);

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "rosie/1.0");
    // Note: no CURLOPT_FAILONERROR — we want to inspect the HTTP code ourselves
    // so callers (like the branch-then-tag fallback) can react to a 404 quietly.

    if (g_verbose) {
        curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
    }

    CURLcode res = curl_easy_perform(curl);

    fclose(fp);

    if (res != CURLE_OK) {
        log_error("Download failed: %s", curl_easy_strerror(res));
        remove(output_path);
        curl_easy_cleanup(curl);
        return -1;
    }

    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
    if (out_http_code) *out_http_code = http_code;

    curl_easy_cleanup(curl);

    if (http_code >= 400) {
        // Caller will decide whether to log; remove the partial file regardless.
        remove(output_path);
        return 0;  // transport ok; HTTP failure is signaled via http_code
    }

    log_debug("Downloaded to: %s", output_path);
    return 0;
}

int download_file(const char *url, const char *output_path) {
    long http_code = 0;
    if (download_file_internal(url, output_path, &http_code) != 0) {
        return -1;
    }
    if (http_code >= 400) {
        log_error("HTTP error: %ld", http_code);
        return -1;
    }
    return 0;
}

int download_package_tarball(const PackageSpec *spec, const char *output_path) {
    if (!spec) return -1;

    // Try as a branch first.
    char *url = build_tarball_url(spec, REF_KIND_BRANCH);
    long http_code = 0;
    int transport = download_file_internal(url, output_path, &http_code);
    spm_free(url);

    if (transport != 0) {
        return -1;  // network/curl error already logged
    }
    if (http_code < 400) {
        return 0;  // success as a branch
    }
    if (http_code != 404) {
        log_error("HTTP error: %ld", http_code);
        return -1;
    }

    // Branch path 404'd — try as a tag.
    log_debug("Ref '%s' not found as branch, trying as tag", spec->ref);
    url = build_tarball_url(spec, REF_KIND_TAG);
    transport = download_file_internal(url, output_path, &http_code);
    spm_free(url);

    if (transport != 0) {
        return -1;
    }
    if (http_code >= 400) {
        log_error("Ref '%s' not found as branch or tag (HTTP %ld)", spec->ref, http_code);
        return -1;
    }
    return 0;
}

File: src/download.rs
// Package-spec parsing + tarball-URL construction + the branch-then-tag
// fallback download. Mirrors download.c.
//
// Supported spec forms:
//   owner/repo[@ref][#skill]           — remote, default ref "main"
//   ./<path>, /<abs>, ~/<rel>, .., .   — local symlinkable skill
//   file://<rel>                       — same, from a lockfile source
//   npm:<pkg>[#<rel-path>]             — npm-ref source (from lockfile)

use crate::os;
use crate::util;
use std::path::{Path, PathBuf};

pub const LOCAL_SOURCE_PREFIX: &str = "file://";
pub const NPM_SOURCE_PREFIX: &str = "npm:";

#[derive(Debug, Clone)]
pub struct PackageSpec {
    pub owner: Option<String>,
    pub repo: Option<String>,
    pub ref_: Option<String>, // defaulted to "main" for remote specs
    pub ref_explicit: bool,
    pub skill_in_spec: Option<String>,
    pub is_local: bool,
    pub local_path: Option<String>,
}

impl PackageSpec {
    pub fn is_remote(&self) -> bool {
        !self.is_local
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RefKind {
    Branch,
    Tag,
}

pub fn source_is_local(source: &str) -> bool {
    source.starts_with(LOCAL_SOURCE_PREFIX)
}

pub fn source_local_path(source: &str) -> Option<&str> {
    source.strip_prefix(LOCAL_SOURCE_PREFIX)
}

pub fn source_is_npm(source: &str) -> bool {
    source.starts_with(NPM_SOURCE_PREFIX)
}

pub fn source_npm_after_prefix(source: &str) -> Option<&str> {
    source.strip_prefix(NPM_SOURCE_PREFIX)
}

/// Split "npm:<pkg>#<file>" into (pkg, file). file is None when no #.
pub fn source_npm_split(source: &str) -> Option<(String, Option<String>)> {
    let body = source_npm_after_prefix(source)?;
    // Last '#' (defensive; npm names contain no '#').
    if let Some(idx) = body.rfind('#') {
        let pkg = body[..idx].to_string();
        let file = if idx + 1 < body.len() {
            Some(body[idx + 1..].to_string())
        } else {
            None
        };
        Some((pkg, file))
    } else {
        Some((body.to_string(), None))
    }
}

/// True if the user-supplied argument is a local path rather than owner/repo.
/// Mirrors `looks_like_local_path` from download.c — leading characters that
/// can't appear in a GitHub owner.
fn looks_like_local_path(spec: &str) -> bool {
    if spec.is_empty() {
        return false;
    }
    if spec == "." {
        return true;
    }
    let b = spec.as_bytes();
    if b[0] == b'/' {
        return true;
    }
    if b.len() >= 2 && b[0] == b'~' && b[1] == b'/' {
        return true;
    }
    if b.len() >= 2 && b[0] == b'.' && b[1] == b'/' {
        return true;
    }
    if b.len() >= 3 && b[0] == b'.' && b[1] == b'.' && b[2] == b'/' {
        return true;
    }
    false
}

/// Resolve a user-supplied path to a "./<rel>" form rooted at the current
/// working directory. Expands a leading `~/`, canonicalizes via std::fs::
/// canonicalize, rejects paths outside the cwd.
fn canonicalize_local_path(user_path: &str) -> Option<String> {
    if user_path.is_empty() {
        return None;
    }

    let expanded: PathBuf = if let Some(rest) = user_path.strip_prefix("~/") {
        match os::home_dir() {
            Some(h) => PathBuf::from(h).join(rest),
            None => {
                crate::log::error("Cannot expand ~ (HOME not set)");
                return None;
            }
        }
    } else {
        PathBuf::from(user_path)
    };

    let abs = match std::fs::canonicalize(&expanded) {
        Ok(p) => p,
        Err(_) => {
            crate::log::error(&format!("Cannot resolve path: {user_path}"));
            return None;
        }
    };

    let cwd = match os::current_dir() {
        Ok(p) => p,
        Err(_) => {
            crate::log::error("Cannot get current directory");
            return None;
        }
    };

    let rel = match abs.strip_prefix(&cwd) {
        Ok(p) => p,
        Err(_) => {
            crate::log::error(&format!(
                "Local skill path is outside the project: {user_path}"
            ));
            return None;
        }
    };

    let rel_str = rel.to_string_lossy();
    if rel_str.is_empty() || rel_str == "" {
        Some(".".to_string())
    } else {
        Some(format!("./{rel_str}"))
    }
}

pub fn parse(spec: &str) -> Option<PackageSpec> {
    // Local-path / file:// shortcut.
    let local_input: Option<&str> = if source_is_local(spec) {
        source_local_path(spec)
    } else if looks_like_local_path(spec) {
        Some(spec)
    } else {
        None
    };
    if let Some(p) = local_input {
        let canonical = canonicalize_local_path(p)?;
        return Some(PackageSpec {
            owner: None,
            repo: None,
            ref_: None,
            ref_explicit: false,
            skill_in_spec: None,
            is_local: true,
            local_path: Some(canonical),
        });
    }

    let work = spec.to_string();

    // @ref suffix
    let (work, ref_, ref_explicit) = match work.split_once('@') {
        Some((head, tail)) => (head.to_string(), tail.to_string(), true),
        None => (work, "main".to_string(), false),
    };

    // #skill suffix (after stripping @ref)
    let (work, skill_in_spec) = match work.split_once('#') {
        Some((head, tail)) if !tail.is_empty() => (head.to_string(), Some(tail.to_string())),
        Some((head, _)) => (head.to_string(), None),
        None => (work, None),
    };

    // owner/repo
    let (owner, repo) = match work.split_once('/') {
        Some((o, r)) => (o.to_string(), r.to_string()),
        None => {
            crate::log::error(&format!("Invalid package spec: {spec} (expected owner/repo)"));
            return None;
        }
    };
    if owner.is_empty() || repo.is_empty() {
        crate::log::error(&format!(
            "Invalid package spec: {spec} (empty owner or repo)"
        ));
        return None;
    }

    Some(PackageSpec {
        owner: Some(owner),
        repo: Some(repo),
        ref_: Some(ref_),
        ref_explicit,
        skill_in_spec,
        is_local: false,
        local_path: None,
    })
}

// ---- URL building ----------------------------------------------------------

pub fn build_tarball_url(spec: &PackageSpec, kind: RefKind) -> Option<String> {
    let owner = spec.owner.as_ref()?;
    let repo = spec.repo.as_ref()?;
    let r = spec.ref_.as_ref()?;
    let kind_segment = match kind {
        RefKind::Tag => "tags",
        RefKind::Branch => "heads",
    };
    let base = crate::http::github_base_url();
    Some(format!(
        "{base}/{owner}/{repo}/archive/refs/{kind_segment}/{r}.tar.gz"
    ))
}

// ---- tarball download with branch-then-tag fallback -----------------------

/// Download the package tarball. Tries refs/heads/<ref> first; on 404
/// falls back to refs/tags/<ref>. Returns 0 on success, -1 on failure.
pub fn download_package_tarball(spec: &PackageSpec, output_path: &Path) -> i32 {
    let url = match build_tarball_url(spec, RefKind::Branch) {
        Some(u) => u,
        None => return -1,
    };
    let status = crate::http::fetch_to_file(&url, output_path);

    if status < 0 {
        return -1;
    }
    if status < 400 {
        return 0;
    }
    if status != 404 {
        crate::log::error(&format!("HTTP error: {status}"));
        return -1;
    }

    let ref_name = spec.ref_.as_deref().unwrap_or("");
    crate::log::debug(&format!(
        "Ref '{ref_name}' not found as branch, trying as tag"
    ));
    let url = match build_tarball_url(spec, RefKind::Tag) {
        Some(u) => u,
        None => return -1,
    };
    let status = crate::http::fetch_to_file(&url, output_path);
    if status < 0 {
        return -1;
    }
    if status >= 400 {
        crate::log::error(&format!(
            "Ref '{ref_name}' not found as branch or tag (HTTP {status})"
        ));
        return -1;
    }
    0
}

/// Build "<base>.tmp" — convenience wrapper used by install.rs.
pub fn tmp_archive_path(dest_dir: &Path, name: &str) -> PathBuf {
    let mut p = dest_dir.to_path_buf();
    p.push(name);
    util::tmp_path_for(&p)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_basic() {
        let s = parse("foo/bar").unwrap();
        assert_eq!(s.owner.as_deref(), Some("foo"));
        assert_eq!(s.repo.as_deref(), Some("bar"));
        assert_eq!(s.ref_.as_deref(), Some("main"));
        assert!(!s.ref_explicit);
    }

    #[test]
    fn parse_pinned() {
        let s = parse("foo/bar@v1.0.0").unwrap();
        assert_eq!(s.ref_.as_deref(), Some("v1.0.0"));
        assert!(s.ref_explicit);
    }

    #[test]
    fn parse_with_skill() {
        let s = parse("foo/bar#my-skill@main").unwrap();
        assert_eq!(s.ref_.as_deref(), Some("main"));
        assert!(s.ref_explicit);
        assert_eq!(s.skill_in_spec.as_deref(), Some("my-skill"));
    }

    #[test]
    fn parse_invalid() {
        crate::log::clear_last_error();
        assert!(parse("foo").is_none());
        assert!(crate::log::last_error_message().is_some());
    }

    #[test]
    fn npm_split() {
        let (p, f) = source_npm_split("npm:react#README.md").unwrap();
        assert_eq!(p, "react");
        assert_eq!(f.as_deref(), Some("README.md"));
        let (p, f) = source_npm_split("npm:@tanstack/react-query#docs/x.md").unwrap();
        assert_eq!(p, "@tanstack/react-query");
        assert_eq!(f.as_deref(), Some("docs/x.md"));
    }

    #[test]
    fn url_build() {
        let spec = parse("vercel/next.js@v14.0.0").unwrap();
        // Use a deterministic base URL for the test.
        std::env::set_var("ROSIE_GITHUB_BASE_URL", "https://example.test");
        assert_eq!(
            build_tarball_url(&spec, RefKind::Tag).unwrap(),
            "https://example.test/vercel/next.js/archive/refs/tags/v14.0.0.tar.gz"
        );
        std::env::remove_var("ROSIE_GITHUB_BASE_URL");
    }
}
